Stop calling `RSpec::Core::Runner#run` multiple times in Queue Mode (#237)
KnapsackPro/knapsack_pro-ruby: CHANGELOG.md

# Changelog

### 7.0.0

* __(breaking change)__ RSpec in Queue Mode:
  * The default for `KNAPSACK_PRO_LOG_LEVEL` is `info` instead of `debug`.
  * The RSpec `before(:suite)` and `after(:suite)` hooks change:

    __Before:__<br>
    The `before(:suite)` and `after(:suite)` hooks are executed multiple times. Each time for a set of tests fetched from Knapsack Pro Queue API.

    __After:__<br>
    The `before(:suite)` and `after(:suite)` hooks are executed only once. The `before(:suite)` hook is executed before starting tests. The `after(:suite)` hook is executed after all tests are completed. It is the default RSpec behavior expected by users.

  * The `KnapsackPro::Hooks::Queue.after_queue` hook change:

    __Before:__<br>
    The `KnapsackPro::Hooks::Queue.after_queue` hook is executed outside of the `after(:suite)` hook.

    __After:__<br>

>     The `KnapsackPro::Hooks::Queue.after_queue` hook is executed __inside__ of the `after(:suite)` hook.

Author: In the past, I really wanted to call `KnapsackPro::Hooks::Queue.after_queue` inside of the `after(:suite)` hook, but it was impossible to know which batch is the final one. You know that after you execute the previous batch. So it is too late to run a hook if all tests are already executed. It must have been called outside of the `after(:suite)` context.

Now, there is no problem like that, and we can do what should be working from the very beginning. The `KnapsackPro::Hooks::Queue.after_queue` hook should be run within the `after(:suite)` context.

I don't think we can add anything useful here. Users simply get now access to the RSpec context within the `KnapsackPro::Hooks::Queue.after_queue` hook, which was not possible before.

* RSpec recommended changes in your project:
  * Remove the following code if you use Queue Mode and the `rspec_junit_formatter` gem to generate JUnit XML or JSON reports. The `KnapsackPro::Hooks::Queue.after_subset_queue` hook is no longer needed to generate a valid XML/JSON report.

    ```ruby

>     # REMOVE THE FOLLOWING CODE

Reviewer: ```suggestion
    # REMOVE THE FOLLOWING CODE

```

    # spec_helper.rb or rails_helper.rb
    TMP_REPORT = "tmp/rspec_#{ENV['KNAPSACK_PRO_CI_NODE_INDEX']}.xml"
    FINAL_REPORT = "tmp/final_rspec_#{ENV['KNAPSACK_PRO_CI_NODE_INDEX']}.xml"

    KnapsackPro::Hooks::Queue.after_subset_queue do |queue_id, subset_queue_id|
      if File.exist?(TMP_REPORT)
        FileUtils.mv(TMP_REPORT, FINAL_REPORT)
      end
    end
    ```

> 
>     You can keep the Knapsack Pro command unchanged with the `--format RspecJunitFormatter --out tmp/rspec.xml` options.
> 
>     ```bash
>     bundle exec rake "knapsack_pro:queue:rspec[--format documentation --format RspecJunitFormatter --out tmp/rspec.xml]"
>     ```

Reviewer: ```suggestion
```

>     Reference in [docs](https://docs.knapsackpro.com/ruby/rspec/#formatters-rspec_junit_formatter-json).
> 
>     See [docs](https://docs.knapsackpro.com/ruby/circleci/) if you use CircleCI.

Reviewer: ```suggestion
    Learn more about [using Knapsack Pro with RSpec formatters](https://docs.knapsackpro.com/ruby/rspec/#formatters-rspec_junit_formatter-json) and [using Knapsack Pro with CircleCI](https://docs.knapsackpro.com/ruby/circleci/) in the docs.
```

>   * Replace the following if you are using Queue Mode and have a version of the `percy-capybara` gem that is older than version `4`:

Reviewer: ```suggestion
  * Replace the following if you are using Queue Mode and have a version of the `percy-capybara` gem that is older than version `4`:
```
```suggestion
  * Replace the following code if you are using Queue Mode and the `percy-capybara` gem on a version older than 4:
```

    Before:

    ```ruby
    KnapsackPro::Hooks::Queue.before_queue { |queue_id| Percy::Capybara.initialize_build }
    KnapsackPro::Hooks::Queue.after_queue { |queue_id| Percy::Capybara.finalize_build }
    ```

    After:

    ```ruby
    # recommended
    before(:suite) { Percy::Capybara.initialize_build }
    after(:suite) { Percy::Capybara.finalize_build }
    ```

>     Reference in [docs](https://docs.knapsackpro.com/ruby/hooks/#percy-capybara) for the latest Percy version.

Reviewer: ```suggestion
    Learn more about [using Knapsack Pro with Percy](https://docs.knapsackpro.com/ruby/rspec/#formatters-rspec_junit_formatter-json) in the docs.
```

Author: done: https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/5341b897b0c7c0bb6fceec63419ec2a84cdc9459 I fixed the wrong url.

>   * We are no longer modifying the default RSpec formatters in Queue Mode. The `KNAPSACK_PRO_MODIFY_DEFAULT_RSPEC_FORMATTERS` environment variable is removed. You can remove it from the CI config if you are using it.

Reviewer: ```suggestion
  * We are no longer modifying the default RSpec formatters in Queue Mode. You can remove the `KNAPSACK_PRO_MODIFY_DEFAULT_RSPEC_FORMATTERS` environment variable from your CI config if you are using it.
```

* RSpec improvements in Queue Mode:

>   * The processing of termination signals - HUP, INT, TERM, ABRT, QUIT, USR1, and USR2 - has been improved for increased speed. Now, the process will terminate promptly before the execution of the next top-level RSpec example group (defined by `describe` or `context` blocks) within a spec file.

Reviewer: ```suggestion
  * Termination signals (`HUP`, `INT`, `TERM`, `ABRT`, `QUIT`, `USR1`, and `USR2`) are handled earlier: the process will terminate before the next top-level example group (`describe` or `context`) instead of waiting for the next Knapsack Pro batch of tests.
```

>     * It uses `1` as a default exit code when the process is terminated.
>     * Or, it uses a custom exit code defined by the `--error-exit-code` option.

Reviewer: Already mentioned below.

```suggestion
```

>   * Respect the `--error-exit-code` option. It overrides the exit code when there are errors outside the test examples that RSpec cannot handle. For example, a lack of memory or a termination signal.

Reviewer: ```suggestion
  * Respect the `--error-exit-code` option. It sets a custom exit code (instead of `1`) when RSpec fails outside an example (e.g. lack of memory, termination signal).
```
